File: examples/node/src/customer.ts
/**
 * Copyright 2026 Chengxi Luo
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * `CustomerRecord` mirrors examples/cases/customer.yaml — a store account.
 *
 * `Address` and `Money` mirror the reusable address.yaml and money.yaml it
 * imports; order.ts reuses both, as the Go worker does.
 *
 * This is the only type in the suite carrying two formats, and the second one
 * is the point: `binary` is a layout written by hand below, `json` goes through
 * `JSON.stringify`, so the two fail in completely different ways. Both declare
 * `oracle: semantic`, so what has to match is the decoded value rather than the
 * bytes — Go's encoder HTML-escapes `<`, `>` and `&` and always escapes
 * U+2028/U+2029, and under a byte oracle every worker would have to reproduce
 * that.
 *
 * It is also the first model here with nested structs, and `model:` on those
 * fields is what the binding needs to hand them back as `Address` and `Money`
 * rather than as raw FieldMaps: TypeScript has erased the property type by the
 * time the value arrives, the same erasure that makes `@Serify.sum` list its
 * arms.
 *
 * Go is the --ref language and owns the layout; see examples/go/wire.go.
 */

import { Serify } from '@chengxilo/serify';

import { lenPrefixed, lenPrefixedStr } from './wire';

/** A cursor over the input buffer, so the readers below stay one-liners. */
export class Reader {
  off = 0;
  constructor(readonly buf: Buffer) {}

  u8(): number { return this.buf[this.off++]; }
  u32(): number { const v = this.buf.readUInt32LE(this.off); this.off += 4; return v; }

  bytes(): Buffer {
    const n = this.u32();
    const b = Buffer.from(this.buf.subarray(this.off, this.off + n)); // copy: a subarray aliases
    this.off += n;
    return b;
  }

  str(): string { return this.bytes().toString('utf8'); }
}

@Serify.Model()
export class Address {
  @Serify.field() recipient = '';
  @Serify.field() street = '';
  @Serify.field() city = '';
  @Serify.field() country = '';
  @Serify.field() postal_code = '';

  /** A struct is its fields back to back, in schema order — nothing frames it. */
  pack(): Buffer {
    return Buffer.concat([
      lenPrefixedStr(this.recipient), lenPrefixedStr(this.street),
      lenPrefixedStr(this.city), lenPrefixedStr(this.country),
      lenPrefixedStr(this.postal_code),
    ]);
  }

  static unpack(r: Reader): Address {
    const a = new Address();
    a.recipient = r.str();
    a.street = r.str();
    a.city = r.str();
    a.country = r.str();
    a.postal_code = r.str();
    return a;
  }
}

@Serify.Model()
export class Money {
  @Serify.field() currency = '';
  @Serify.field() amount_minor = 0n;

  pack(): Buffer {
    const amt = Buffer.alloc(8);
    amt.writeBigInt64LE(this.amount_minor, 0);
    return Buffer.concat([lenPrefixedStr(this.currency), amt]);
  }

  static unpack(r: Reader): Money {
    const m = new Money();
    m.currency = r.str();
    m.amount_minor = r.buf.readBigInt64LE(r.off);
    r.off += 8;
    return m;
  }
}

@Serify.Model()
export class CustomerRecord {
  @Serify.field() customer_id = 0n;
  @Serify.field() email = '';
  @Serify.field() display_name = '';
  @Serify.field() age = 0;
  @Serify.field() email_verified = false;
  @Serify.field() fraud_score = 0;
  @Serify.field() loyalty_points = 0;
  @Serify.field() signup_ts = 0n;
  @Serify.field() avatar_sha256: Buffer = Buffer.alloc(0);
  @Serify.field() pin: number[] = [0, 0, 0, 0];
  @Serify.field() referral_code: string | null = null;
  @Serify.field({ model: Money }) store_credit = new Money();
  @Serify.field({ model: Address }) shipping_addresses: Address[] = [];
  @Serify.field({ model: Address }) address_book = new Map<string, Address>();
  @Serify.field() wishlist_skus: string[] = [];
  @Serify.field() preferences = new Map<string, string>();

  marshal(): Buffer {
    const head = Buffer.alloc(8);
    head.writeBigUInt64LE(this.customer_id, 0);

    const scalars = Buffer.alloc(18);
    scalars.writeUInt8(this.age, 0);
    scalars.writeUInt8(this.email_verified ? 1 : 0, 1);
    scalars.writeFloatLE(this.fraud_score, 2);
    scalars.writeUInt32LE(this.loyalty_points, 6);
    scalars.writeBigInt64LE(this.signup_ts, 10);

    const parts: Buffer[] = [
      head,
      lenPrefixedStr(this.email),
      lenPrefixedStr(this.display_name),
      scalars,
      lenPrefixed(this.avatar_sha256),
      // array<T,N> carries no count: N is fixed by the schema.
      Buffer.from(this.pin),
    ];

    // optional<string>: a presence flag, then the value if present. An empty
    // string is present, which is why the flag cannot be inferred from it.
    if (this.referral_code === null) parts.push(Buffer.from([0]));
    else parts.push(Buffer.from([1]), lenPrefixedStr(this.referral_code));

    parts.push(this.store_credit.pack());

    parts.push(count(this.shipping_addresses.length));
    for (const a of this.shipping_addresses) parts.push(a.pack());

    // Entry order is the Map's own — deliberately not sorted. A map is
    // unordered, so customer declares `oracle: semantic` and the decoded value
    // is what gets compared. See docs/protocol.md.
    parts.push(count(this.address_book.size));
    for (const [k, a] of this.address_book) parts.push(lenPrefixedStr(k), a.pack());

    parts.push(count(this.wishlist_skus.length));
    for (const s of this.wishlist_skus) parts.push(lenPrefixedStr(s));

    parts.push(count(this.preferences.size));
    for (const [k, v] of this.preferences) parts.push(lenPrefixedStr(k), lenPrefixedStr(v));

    return Buffer.concat(parts);
  }

  static unmarshal(data: Buffer): CustomerRecord {
    const c = new CustomerRecord();
    const r = new Reader(data);

    c.customer_id = data.readBigUInt64LE(0);
    r.off = 8;
    c.email = r.str();
    c.display_name = r.str();

    c.age = data.readUInt8(r.off);
    c.email_verified = data.readUInt8(r.off + 1) !== 0;
    c.fraud_score = data.readFloatLE(r.off + 2);
    c.loyalty_points = data.readUInt32LE(r.off + 6);
    c.signup_ts = data.readBigInt64LE(r.off + 10);
    r.off += 18;

    c.avatar_sha256 = r.bytes();

    c.pin = [...data.subarray(r.off, r.off + 4)];
    r.off += 4;

    c.referral_code = r.u8() === 0 ? null : r.str();

    c.store_credit = Money.unpack(r);

    c.shipping_addresses = [];
    for (let n = r.u32(); n > 0; n--) c.shipping_addresses.push(Address.unpack(r));

    c.address_book = new Map();
    for (let n = r.u32(); n > 0; n--) c.address_book.set(r.str(), Address.unpack(r));

    c.wishlist_skus = [];
    for (let n = r.u32(); n > 0; n--) c.wishlist_skus.push(r.str());

    c.preferences = new Map();
    for (let n = r.u32(); n > 0; n--) c.preferences.set(r.str(), r.str());

    return c;
  }

  /**
   * `bytes` is base64 in JSON, and the 64-bit fields are plain numbers: that is
   * what the reference worker's `[]byte` and `uint64` marshal to, and the
   * semantic oracle decodes our output with it.
   */
  toJSON(): Buffer {
    const obj = {
      customer_id: this.customer_id.toString(),
      email: this.email,
      display_name: this.display_name,
      age: this.age,
      email_verified: this.email_verified,
      fraud_score: this.fraud_score,
      loyalty_points: this.loyalty_points,
      signup_ts: this.signup_ts.toString(),
      avatar_sha256: this.avatar_sha256.toString('base64'),
      pin: this.pin,
      referral_code: this.referral_code,
      store_credit: {
        currency: this.store_credit.currency,
        amount_minor: this.store_credit.amount_minor.toString(),
      },
      shipping_addresses: this.shipping_addresses.map(addrJSON),
      address_book: Object.fromEntries([...this.address_book].map(([k, a]) => [k, addrJSON(a)])),
      wishlist_skus: this.wishlist_skus,
      preferences: Object.fromEntries(this.preferences),
    };
    return Buffer.from(unquoteBigints(JSON.stringify(obj)), 'utf8');
  }

  static fromJSON(data: Buffer): CustomerRecord {
    const c = new CustomerRecord();
    const o = JSON.parse(quoteBigints(data.toString('utf8')));

    c.customer_id = BigInt(o.customer_id);
    c.email = o.email;
    c.display_name = o.display_name;
    c.age = o.age;
    c.email_verified = o.email_verified;
    // A JSON number is a double; narrow it the way the wire does, so a float32
    // field holds a value float32 can actually represent.
    c.fraud_score = Math.fround(o.fraud_score);
    c.loyalty_points = o.loyalty_points;
    c.signup_ts = BigInt(o.signup_ts);
    c.avatar_sha256 = Buffer.from(o.avatar_sha256 ?? '', 'base64');
    c.pin = o.pin;
    c.referral_code = o.referral_code ?? null;
    c.store_credit = moneyFromJSON(o.store_credit);
    c.shipping_addresses = (o.shipping_addresses ?? []).map(addrFromJSON);
    c.address_book = new Map(
      Object.entries(o.address_book ?? {}).map(([k, v]) => [k, addrFromJSON(v)]));
    c.wishlist_skus = o.wishlist_skus ?? [];
    c.preferences = new Map(Object.entries(o.preferences ?? {}));
    return c;
  }
}

function count(n: number): Buffer {
  const b = Buffer.alloc(4);
  b.writeUInt32LE(n, 0);
  return b;
}

// 64-bit integers in JSON
//
// JSON's only number is the double, and customer's boundary case does not fit
// in one: max uint64 rounds up to 2^64 and comes back out of range. JSON.parse
// offers no way to see the undamaged token -- Node 22's JSON.rawJSON does, but
// CI builds on 20 -- so these three fields cross the text boundary as quoted
// strings, unquoted on the way out and requoted on the way in. They are the
// only 64-bit fields customer has.
//
// This rewrites the text rather than the parse tree, so a *string value*
// spelling one of these keys followed by a colon and digits would be lifted
// too. No case produces one, and the escape-hunting cases cannot: a `"` inside
// a JSON string is written `\"`.
const BIG_KEYS = 'customer_id|signup_ts|amount_minor';

function unquoteBigints(text: string): string {
  return text.replace(new RegExp(`"(${BIG_KEYS})":"(-?\\d+)"`, 'g'), '"$1":$2');
}

function quoteBigints(text: string): string {
  return text.replace(new RegExp(`"(${BIG_KEYS})":(-?\\d+)`, 'g'), '"$1":"$2"');
}

function addrJSON(a: Address) {
  return {
    recipient: a.recipient, street: a.street, city: a.city,
    country: a.country, postal_code: a.postal_code,
  };
}

function addrFromJSON(o: any): Address {
  const a = new Address();
  a.recipient = o.recipient;
  a.street = o.street;
  a.city = o.city;
  a.country = o.country;
  a.postal_code = o.postal_code;
  return a;
}

function moneyFromJSON(o: any): Money {
  const m = new Money();
  m.currency = o.currency;
  m.amount_minor = BigInt(o.amount_minor);
  return m;
}

File: examples/python/customer.py
# Copyright 2026 Chengxi Luo
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""`CustomerRecord` mirrors examples/cases/customer.yaml — a store account.

`Address` and `Money` mirror the reusable address.yaml and money.yaml it
imports; order.py reuses both, as the Go worker does.

This is the only type in the suite carrying two formats, and the second one is
the point: `binary` is a layout this file writes by hand, while `json` goes
through the stdlib encoder, so the two exercise completely different failure
modes. Both declare `oracle: semantic`, so what has to match the reference is
the decoded value, not the bytes — which is what makes a second format
affordable at all. Go's encoder HTML-escapes `<`, `>` and `&` and always
escapes U+2028/U+2029; no other language does, and under a byte oracle every
worker would have to reproduce that quirk.

`ensure_ascii=True` (the json default) is deliberate here: escaping every
non-ASCII character to \\uXXXX sidesteps the question of whose escaping rules
win, and Go's decoder reads the escapes back — including the surrogate pairs the
emoji in the `unicode` case turn into.

Go is the --ref language and owns the byte layout; see examples/go/wire.go.
"""

import base64
import json
import struct
from dataclasses import dataclass

from serify import serify_model


def pack_str(s: str) -> bytes:
    """u32 byte length, then the UTF-8 bytes. Public because order.py imports it
    along with Address and Money."""
    raw = s.encode('utf-8')
    return struct.pack('<I', len(raw)) + raw


def unpack_str(data: bytes, pos: int):
    n = struct.unpack_from('<I', data, pos)[0]
    pos += 4
    return data[pos:pos + n].decode('utf-8'), pos + n


@serify_model
@dataclass
class Address:
    recipient: str
    street: str
    city: str
    country: str
    postal_code: str

    def pack(self) -> bytes:
        """A struct is its fields back to back, in schema order — no count, no
        length, nothing framing it."""
        return (pack_str(self.recipient) + pack_str(self.street)
                + pack_str(self.city) + pack_str(self.country)
                + pack_str(self.postal_code))

    @classmethod
    def unpack(cls, data: bytes, pos: int):
        recipient, pos = unpack_str(data, pos)
        street, pos = unpack_str(data, pos)
        city, pos = unpack_str(data, pos)
        country, pos = unpack_str(data, pos)
        postal_code, pos = unpack_str(data, pos)
        return cls(recipient, street, city, country, postal_code), pos


@serify_model
@dataclass
class Money:
    currency: str
    amount_minor: int

    def pack(self) -> bytes:
        return pack_str(self.currency) + struct.pack('<q', self.amount_minor)

    @classmethod
    def unpack(cls, data: bytes, pos: int):
        currency, pos = unpack_str(data, pos)
        amount_minor = struct.unpack_from('<q', data, pos)[0]
        return cls(currency, amount_minor), pos + 8


@serify_model
@dataclass
class CustomerRecord:
    customer_id: int
    email: str
    display_name: str
    age: int
    email_verified: bool
    fraud_score: float
    loyalty_points: int
    signup_ts: int
    avatar_sha256: bytes
    pin: "list[int]"
    referral_code: "str | None"
    store_credit: Money
    shipping_addresses: "list[Address]"
    address_book: "dict[str, Address]"
    wishlist_skus: "list[str]"
    preferences: "dict[str, str]"

    def marshal(self) -> bytes:
        buf = bytearray()
        buf += struct.pack('<Q', self.customer_id)
        buf += pack_str(self.email)
        buf += pack_str(self.display_name)
        buf += struct.pack('<BB', self.age, 1 if self.email_verified else 0)
        buf += struct.pack('<fIq', self.fraud_score, self.loyalty_points, self.signup_ts)

        buf += struct.pack('<I', len(self.avatar_sha256)) + self.avatar_sha256

        # array<T,N> carries no count: N is fixed by the schema.
        buf += bytes(self.pin)

        # optional<string>: a presence flag, then the value if present. An empty
        # string is present, which is why the flag cannot be inferred from it.
        if self.referral_code is None:
            buf += struct.pack('<B', 0)
        else:
            buf += struct.pack('<B', 1) + pack_str(self.referral_code)

        buf += self.store_credit.pack()

        buf += struct.pack('<I', len(self.shipping_addresses))
        for a in self.shipping_addresses:
            buf += a.pack()

        # Entry order is the dict's own — deliberately not sorted. A map is
        # unordered, so customer declares `oracle: semantic` and the decoded
        # value is what gets compared. See docs/protocol.md.
        buf += struct.pack('<I', len(self.address_book))
        for k, a in self.address_book.items():
            buf += pack_str(k) + a.pack()

        buf += struct.pack('<I', len(self.wishlist_skus))
        for s in self.wishlist_skus:
            buf += pack_str(s)

        buf += struct.pack('<I', len(self.preferences))
        for k, v in self.preferences.items():
            buf += pack_str(k) + pack_str(v)

        return bytes(buf)

    @classmethod
    def unmarshal(cls, data: bytes) -> "CustomerRecord":
        pos = 0
        customer_id = struct.unpack_from('<Q', data, pos)[0]
        pos += 8

        email, pos = unpack_str(data, pos)
        display_name, pos = unpack_str(data, pos)

        age, verified, fraud_score, loyalty_points, signup_ts = struct.unpack_from('<BBfIq', data, pos)
        pos += 2 + 4 + 4 + 8

        n = struct.unpack_from('<I', data, pos)[0]
        pos += 4
        avatar_sha256 = bytes(data[pos:pos + n])
        pos += n

        pin = list(data[pos:pos + 4])
        pos += 4

        if data[pos] == 0:
            referral_code = None
            pos += 1
        else:
            referral_code, pos = unpack_str(data, pos + 1)

        store_credit, pos = Money.unpack(data, pos)

        n = struct.unpack_from('<I', data, pos)[0]
        pos += 4
        shipping_addresses = []
        for _ in range(n):
            a, pos = Address.unpack(data, pos)
            shipping_addresses.append(a)

        n = struct.unpack_from('<I', data, pos)[0]
        pos += 4
        address_book = {}
        for _ in range(n):
            k, pos = unpack_str(data, pos)
            address_book[k], pos = Address.unpack(data, pos)

        n = struct.unpack_from('<I', data, pos)[0]
        pos += 4
        wishlist_skus = []
        for _ in range(n):
            s, pos = unpack_str(data, pos)
            wishlist_skus.append(s)

        n = struct.unpack_from('<I', data, pos)[0]
        pos += 4
        preferences = {}
        for _ in range(n):
            k, pos = unpack_str(data, pos)
            preferences[k], pos = unpack_str(data, pos)

        return cls(
            customer_id=customer_id, email=email, display_name=display_name,
            age=age, email_verified=verified != 0, fraud_score=fraud_score,
            loyalty_points=loyalty_points, signup_ts=signup_ts,
            avatar_sha256=avatar_sha256, pin=pin, referral_code=referral_code,
            store_credit=store_credit, shipping_addresses=shipping_addresses,
            address_book=address_book, wishlist_skus=wishlist_skus,
            preferences=preferences,
        )

    def to_json(self) -> bytes:
        """`bytes` is base64 in JSON, because that is what the reference worker's
        `[]byte` marshals to and the semantic oracle decodes our output with it."""
        obj = {
            "customer_id": self.customer_id,
            "email": self.email,
            "display_name": self.display_name,
            "age": self.age,
            "email_verified": self.email_verified,
            "fraud_score": self.fraud_score,
            "loyalty_points": self.loyalty_points,
            "signup_ts": self.signup_ts,
            "avatar_sha256": base64.b64encode(self.avatar_sha256).decode('ascii'),
            "pin": list(self.pin),
            "referral_code": self.referral_code,
            "store_credit": {
                "currency": self.store_credit.currency,
                "amount_minor": self.store_credit.amount_minor,
            },
            "shipping_addresses": [_addr_json(a) for a in self.shipping_addresses],
            "address_book": {k: _addr_json(a) for k, a in self.address_book.items()},
            "wishlist_skus": list(self.wishlist_skus),
            "preferences": dict(self.preferences),
        }
        return json.dumps(obj, separators=(',', ':')).encode('ascii')

    @classmethod
    def from_json(cls, data: bytes) -> "CustomerRecord":
        o = json.loads(data.decode('utf-8'))
        return cls(
            customer_id=o["customer_id"],
            email=o["email"],
            display_name=o["display_name"],
            age=o["age"],
            email_verified=o["email_verified"],
            # A JSON number is a double; narrow it the way the wire does, so a
            # float32 field holds a value float32 can actually represent.
            fraud_score=struct.unpack('<f', struct.pack('<f', o["fraud_score"]))[0],
            loyalty_points=o["loyalty_points"],
            signup_ts=o["signup_ts"],
            avatar_sha256=base64.b64decode(o["avatar_sha256"] or ""),
            pin=list(o["pin"]),
            referral_code=o["referral_code"],
            store_credit=Money(o["store_credit"]["currency"], o["store_credit"]["amount_minor"]),
            shipping_addresses=[_addr_from_json(a) for a in (o["shipping_addresses"] or [])],
            address_book={k: _addr_from_json(a) for k, a in (o["address_book"] or {}).items()},
            wishlist_skus=list(o["wishlist_skus"] or []),
            preferences=dict(o["preferences"] or {}),
        )


def _addr_json(a: Address) -> dict:
    return {
        "recipient": a.recipient, "street": a.street, "city": a.city,
        "country": a.country, "postal_code": a.postal_code,
    }


def _addr_from_json(o: dict) -> Address:
    return Address(o["recipient"], o["street"], o["city"], o["country"], o["postal_code"])
